### ISSUE TITLE
Incorrect use of SSL_set_tlsext_host_name

### ISSUE THREAD

**Issue by yurivict** [proxytunnel/proxytunnel#69]
[Here](https://github.com/proxytunnel/proxytunnel/blob/master/ptstream.c#L344) the ```SSL_set_tlsext_host_name```'s return value is checked for being equal to SSL_TLSEXT_ERR_OK (=0) as a sign of success, while the manpage says that it returns 0 or 1 and 1 means success:
```
RETURN VALUES
       SSL_CTX_set_tlsext_servername_callback() and
       SSL_CTX_set_tlsext_servername_arg() both always return 1 indicating
       success.  SSL_set_tlsext_host_name() returns 1 on success, 0 in case of
       error.
```

**Comment by ericchansen** [comment]
I am seeing this error in my connection. Related?

```
...
debug3: w32_getpeername ERROR: not sock :2
debug1: identity file C:\\Users\\ericc/.ssh/id_rsa type 0
debug3: Failed to open file:C:/Users/ericc/.ssh/id_rsa-cert error:2
debug3: Failed to open file:C:/Users/ericc/.ssh/id_rsa-cert.pub error:2
debug3: failed to open file:C:/Users/ericc/.ssh/id_rsa-cert error:2
debug1: identity file C:\\Users\\ericc/.ssh/id_rsa-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_for_Windows_8.6
SSL_set_tlsext_host_name returned: 1 (0x1). TLS SNI error, giving up
kex_exchange_identification: Connection closed by remote host
Connection closed by UNKNOWN port 65535
```

**Comment by yurivict** [comment]
Yes, this becomes a fatal error on Windows.


You need to make it non-fatal, after which proxytunnel should work fine.

**Comment by ericchansen** [comment]
I skipped that error.
```
debug1: C:\\Users\\ericc/.ssh/config line 5: Applying options for *
debug3: Failed to open file:C:/ProgramData/ssh/ssh_config error:2
debug3: expanded UserKnownHostsFile '~/.ssh/known_hosts' -> 'C:\\Users\\ericc/.ssh/known_hosts'
debug3: expanded UserKnownHostsFile '~/.ssh/known_hosts2' -> 'C:\\Users\\ericc/.ssh/known_hosts2'
debug1: Authenticator provider $SSH_SK_PROVIDER did not resolve; disabling
debug1: Executing proxy command: exec C:\\msys64\\home\\ericc\\repos\\proxytunnel\\bin\\proxytunnel.exe -E -p somewhere:443 -d 127.0.0.1:22
debug3: spawning "C:\\msys64\\home\\ericc\\repos\\proxytunnel\\bin\\proxytunnel.exe" -E -p somewhere:443 -d 127.0.0.1:22 as subprocess
debug3: w32_getpeername ERROR: not sock :2
debug1: identity file C:\\Users\\ericc/.ssh/id_rsa type 0
debug3: Failed to open file:C:/Users/ericc/.ssh/id_rsa-cert error:2
debug3: Failed to open file:C:/Users/ericc/.ssh/id_rsa-cert.pub error:2
debug3: failed to open file:C:/Users/ericc/.ssh/id_rsa-cert error:2
debug1: identity file C:\\Users\\ericc/.ssh/id_rsa-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_for_Windows_8.6
Via somewhere:443 -> 127.0.0.1:22
SSL_set_tlsext_host_name returned: 1 (0x1). TLS SNI error, BUT ERIC CHANGED THE CODE TO NOT EXIT LET'S GOOO
Certificate verification failed (unable to get local issuer certificate)
kex_exchange_identification: Connection closed by remote host
Connection closed by UNKNOWN port 65535
```
Any ideas?

**Comment by yurivict** [comment]
Ah ... you need to disable certificate verification.

**Comment by ericchansen** [comment]
Thanks for the feedback!

I feel like everything is properly configured because I can connect to it using proxytunnel on Unix based systems without issues. A box with Ubuntu 22.04.2 LTS, a Mac with the latest updates and Ubuntu 22.04.2 LTS in WSL2 all work.

Here is a simplified version of my config on Unix (not including all the extra .conf files).
```
Host somewhere
  HostName somewhere
  ProxyCommand proxytunnel -C /etc/ssl/cert.pem -q -E -p somewhere:443 -d 127.0.0.1:22
  Port 443
Host *
  AddKeysToAgent yes
  IdentityFile ~/.ssh/id_rsa
  ServerAliveInterval 100
  StrictHostKeyChecking no
  User ehansen
```

Here is what it looks like on Windows (using proxytunnel.exe and removed the Unix specific cert part).
```
Host somewhere
  HostName somewhere
  ProxyCommand C:\msys64\home\ericc\repos\proxytunnel\bin\proxytunnel.exe -E -p somewhere:443 -d 127.0.0.1:22
  Port 443
Host *
  AddKeysToAgent yes
  IdentityFile ~/.ssh/id_rsa
  ServerAliveInterval 100
  StrictHostKeyChecking no
  User ehansen
```
I'm not sure how to check with Telnet.

I get a web page that shows this when I try to connect to somwhere:443 with a browser.
```
Bad Request
Your browser sent a request that this server could not understand.
Reason: You're speaking plain HTTP to an SSL-enabled server port.
Instead use the HTTPS scheme to access this URL, please.
Apache/2.4.41 (Ubuntu) Server at somewhere Port 80
```

I built proxytunnel following the msys directions here https://github.com/proxytunnel/proxytunnel/blob/master/INSTALL.md#msys2.

I will try to figure out how to build with Cygwin. I'm not sure how to "into nginx with the HTTP_PROXY_CONNECT port option set to ON running on FreeBSD".

**Comment by yurivict** [comment]
> Here is a simplified version of my config on Unix (not including all the extra .conf files).
> 
> Host somewhere
>   HostName somewhere
>   ProxyCommand proxytunnel -C /etc/ssl/cert.pem -q -E -p somewhere:443 -d 127.0.0.1:22
>   Port 443

Try replacing ```-C /etc/ssl/cert.pem``` with ```--no-check-certificate```.

If you are trying to circumvent a corporate network middlebox - it might substitute certificates and have a corporate certificate authority installed on your corporate device (laptop). It basically breaks into all https connections so your certificate isn't what is seen by proxychannel.

For record, my setup is:
```
host home
        ProxyCommand {somedir}/proxytunnel --no-check-certificate -E -p {proxy-host}:{proxy-port} -d 127.0.0.1:{local-ssh-port-on-the-server}
```

> I'm not sure how to "into nginx with the HTTP_PROXY_CONNECT port option set to ON running on FreeBSD".

I use FreeBSD as server with nginx port that has the option ```HTTP_PROXY_CONNECT=ON``` set.
If you use some other distro - you need to manually add the http_proxy_connecton to nginx.

**Comment by ericchansen** [comment]
That works! Thank you! For anyone else who might see this thread, adding `--no-check-certificate` works.

**Comment by 68420948** [comment]
This issue about _Incorrect use of SSL_set_tlsext_host_name_ is a duplicate of issue #59 and had been fixed with PR #75.

**Comment by 68420948** [comment]
Fix released with proxytunnel v1.11.1
